feat/ create modal component

Okazaki92 opened the pull request:
This modal component contains `children, isModalOpen, setIsModalOpen` Props which are used to open this modal in desired component. To open it, you must create a state to handle modal opening and put some data inside modal.

Ex. 
```
<div>
<button onClick={() => setIsModalOpen(true)}>BUTTON</button>
<Modal isModalOpen={isModalOpen} setIsModalOpen={setIsModalOpen}>
     <p>Modal content</p>
</Modal>
</div>
```

I've also added ESC key event handler for closing window.

NacomiTagiera commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,41 @@
+  return (
+    isModalOpen && (
+      <div
+        onClick={() => setIsModalOpen(false)}
```
Clicking on the modal's backdrop closes the modal. However, this will also close the modal if the user clicks inside the modal itself.

Okazaki92 replied:
I've added `onClick={(e) => e.stopPropagation()}` in child component to prevent closing modal by clicking inside modal itself

NacomiTagiera commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,41 @@
+
+  return (
+    isModalOpen && (
+      <div
```
To make the modal accessible, include ARIA attributes like `role="dialog"` and `aria-modal="true"`.

NacomiTagiera commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,41 @@
+};
+
+const Modal = ({ children, isModalOpen, setIsModalOpen }: ModalProps) => {
+  useEffect(() => {
```
When the modal is open, you might want to prevent the main page from scrolling.
`
useEffect(() => {
  if (isModalOpen) {
    document.body.style.overflow = 'hidden';
    return () => {
      document.body.style.overflow = 'unset';
    };
  }
}, [isModalOpen]);
`

NacomiTagiera commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,41 @@
+import ModalCloseIcon from "@/icons/ModalCloseIcon";
+import { useEffect } from "react";
+
+type ModalProps = {
+  isModalOpen: boolean;
+  setIsModalOpen: (isOpen: boolean) => void;
+  children: React.ReactNode;
+};
```
Consider adding optional customization props for more flexibility

NacomiTagiera commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,41 @@
+
+const Modal = ({ children, isModalOpen, setIsModalOpen }: ModalProps) => {
+  useEffect(() => {
+    const handleKeyDown = (e: KeyboardEvent) => {
+      if (e.key === "Escape") {
+        setIsModalOpen(false);
+      }
```
The current implementation re-renders on every keydown. Consider using useCallback.

NacomiTagiera commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,41 @@
+        className="fixed inset-0 z-50 flex h-screen items-center justify-center overflow-y-scroll bg-slate-900/20 backdrop-blur"
+      >
+        <div className="relative m-auto flex animate-modalAnimation cursor-default flex-col items-center rounded bg-greenMain px-3 pt-12 text-white ease-in desktop:px-10 desktop:pt-[60px]">
+          <button
+            className="absolute right-3 top-3"
+            onClick={() => setIsModalOpen(false)}
+          >
+            <ModalCloseIcon />
+          </button>
```
To make the close button more accessible, include `aria-label="Close modal"` or `<span className="sr-only">Close modal</span>` inside `<button>` tag.

Okazaki92 commented:
1.I've added Modal.test.tsx and Modal.stories.tsx to Modal component
2.Added ARIA attributes and make changes in Modal component according code review from @NacomiTagiera

NacomiTagiera commented on src/components/Modal/Modal.test.tsx:
```diff
@@ -0,0 +1,55 @@
+import { render, screen, fireEvent } from "@testing-library/react";
+import { Modal } from "./Modal";
+
+describe("Modal component", () => {
```
You may also add a test to ensure focus moves to the modal when opened, test if the focus returns to the triggering element when the modal closes, include a snapshot test to ensure the modal's UI does not unintentionally change, test the scroll lock behavior, write a test for event propagation and add a test to check if the modal doesn't close when user clicks any other key than `Escape`.

NacomiTagiera commented on src/components/Modal/Modal.stories.tsx:
```diff
@@ -0,0 +1,40 @@
+import { StoryFn, Meta } from "@storybook/react";
+import { Modal, ModalProps } from "./Modal";
+
+export default {
+  title: "Components/Modal",
+  component: Modal,
+  argTypes: {
+    isModalOpen: {
+      control: {
+        type: "boolean",
+      },
+    },
+    setIsModalOpen: {
+      action: "setIsModalOpen",
+    },
+  },
+} as Meta;
```
The `Storybook` stories for your `Modal` component are good, I can suggest some improvements to make them more comprehensive and useful.
```
import { StoryFn, Meta } from "@storybook/react";
import { useState } from 'react';
import { Modal, ModalProps } from "./Modal";

export default {
  title: "Components/Modal",
  component: Modal,
  parameters: {
    layout: 'centered',
    // Disable default padding to show modal overlay properly
    layout: 'fullscreen',
    // Add docs about the component
    docs: {
      description: {
        component: 'A reusable modal component that supports custom content, keyboard navigation, and click-outside behavior.'
      }
    }
  },
  // Adding more detailed control descriptions
  argTypes: {
    isModalOpen: {
      control: 'boolean',
      description: 'Controls the visibility of the modal'
    },
    setIsModalOpen: {
      action: 'setIsModalOpen',
      description: 'Callback function to control modal state'
    },
    children: {
      control: 'text',
      description: 'Content to be rendered inside the modal'
    }
  }
} as Meta;

// Interactive template that maintains its own state
const InteractiveTemplate: StoryFn<ModalProps> = (args) => {
  const [isOpen, setIsOpen] = useState(false);
  
  return (
    <div>
      <button onClick={() => setIsOpen(true)}>Open Modal</button>
      <Modal 
        {...args} 
        isModalOpen={isOpen}
        setIsModalOpen={setIsOpen}
      />
    </div>
  );
};

// Basic template for static stories
const StaticTemplate: StoryFn<ModalProps> = (args) => <Modal {...args} />;

export const Interactive = InteractiveTemplate.bind({});
Interactive.args = {
  children: <div className="p-4">
    <h2 className="text-xl mb-4">Interactive Modal</h2>
    <p>This modal can be opened and closed with state management</p>
  </div>
};

export const Default = StaticTemplate.bind({});
Default.args = {
  isModalOpen: true,
  children: <div className="p-4">
    <h2 className="text-xl mb-4">Default Modal</h2>
    <p>This is a basic modal example</p>
  </div>
};

export const WithCustomContent = StaticTemplate.bind({});
WithCustomContent.args = {
  isModalOpen: true,
  children: (
    <div className="p-8">
      <h1 className="text-2xl mb-4">Custom Content Modal</h1>
      <p className="mb-4">This modal demonstrates custom styling and content</p>
      <button className="px-4 py-2 bg-blue-500 text-white rounded">
        Action Button
      </button>
    </div>
  )
};

export const LongContent = StaticTemplate.bind({});
LongContent.args = {
  isModalOpen: true,
  children: (
    <div className="p-4">
      <h2 className="text-xl mb-4">Scrollable Content</h2>
      {Array(10).fill(null).map((_, i) => (
        <p key={i} className="mb-4">
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
          Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
        </p>
      ))}
    </div>
  )
};
```

NacomiTagiera reviewed (changes requested):
Your refactored modal component and accompanying tests are solid. They cover the core functionality and edge cases well, ensuring the component behaves as expected. I have only some suggestions for tests and storybook stories.

Okazaki92 commented:
I've added focus on modal open and close and all tests from review
